discovery/watcher.go

```diff
@@ -0,0 +1,367 @@
+package discovery
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/hashicorp/go-hclog"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/hashicorp/consul-server-connection-manager/internal/consul-proto/pbdataplane"
+)
+
+// InitState is the info a caller wants to know after initialization.
+type InitState struct {
+	// GRPCConn is the gRPC connection shared with this library. Use
+	// this to create your gRPC clients. The gRPC connection is
+	// automatically updated to switch to a new server, so you can
+	// use this connection for the lifetime of the associated
+	// Watcher.
+	GRPCConn grpc.ClientConnInterface
+
+	// Token is the ACL token obtain from logging in (if applicable).
+	// If login is not supported this will be set to the static token
+	// from the Config object.
+	Token string
+
+	// Address is the address of current the Consul server the Watcher is using.
+	Address Addr
+
+	// DataplaneFeatures contains the dataplane features supported by the
+	// current Consul server.
+	DataplaneFeatures map[string]bool
+}
+
+// SubscribeState is the info callers need to know when the Watcher's
+// current server changes.
+// TODO: Fill this in.
+type SubscribeState struct{}
+
+type Watcher struct {
+	ctx       context.Context
+	ctxCancel context.CancelFunc
+
+	config Config
+	log    hclog.Logger
+
+	currentServer atomic.Value
+
+	initComplete *event
+	runComplete  *event
+	runOnce      sync.Once
+
+	backoff  backoff.BackOff
+	conn     *grpc.ClientConn
+	resolver *watcherResolver
+	token    string
+
+	// interface to allow us to inject custom server ports for tests
+	discoverer Discoverer
+}
+
+type serverState struct {
+	addr              Addr
+	dataplaneFeatures map[string]bool
+}
+
+func NewWatcher(ctx context.Context, config Config, log hclog.Logger) (*Watcher, error) {
+	if log == nil {
+		log = hclog.NewNullLogger()
+	}
+
+	// TODO: config for backoff values
+	backoff := backoff.NewExponentialBackOff()
+	backoff.MaxElapsedTime = 0 // Allow backing off forever.
+
+	w := &Watcher{
+		config:       config,
+		log:          log,
+		backoff:      backoff,
+		resolver:     newResolver(log),
+		discoverer:   NewNetaddrsDiscoverer(config, log),
+		initComplete: newEvent(),
+		runComplete:  newEvent(),
+	}
+
+	w.ctx, w.ctxCancel = context.WithCancel(ctx)
+	w.currentServer.Store(serverState{})
+
+	var cred credentials.TransportCredentials
+	if tls := w.config.TLS; tls != nil {
+		cred = credentials.NewTLS(tls)
+	} else {
+		cred = insecure.NewCredentials()
+	}
+	dialOpts := []grpc.DialOption{
+		grpc.WithTransportCredentials(cred),
+		grpc.WithResolvers(w.resolver), // note: experimental api.
```

Comment:
This looks neat, but also concerning that it is an experimental API. 
If we cannot use this API for some reason in the future (in case it is removed), what's the alternative?  The new grpc connection would also be published in the subscribed state I guess?

Reply:
Yeah, we could always fall back to notifying users when there's a new gRPC connection object via the subscribe interface. 

Or we could potentially return our own implementation of a `grpc.ClientConnInterface` that switches the connection automatically.

Reply:
We depend on this experimental API in our cloud services. There was a breaking change in a minor release (around retrying resolution) that was a near-miss for us. Not an argument for or against, just calling it out!

Reply:
FWIW: I think we should stick with it!